nomad/resource_csi_volume.go

```diff
@@ -336,13 +335,77 @@ func resourceCSIVolume() *schema.Resource {
 					},
 				},
 			},
+			"external_id": {
+				Description: "The ID of the physical volume from the storage provider.",
+				Computed:    true,
+				Type:        schema.TypeString,
+			},
 		},
 	}
 }
 
-func resourceCSIVolumeRead(d *schema.ResourceData, meta any) error {
-	_, err := readCSIVolume(d, meta)
-	return err
+// resourceCSIVolumeRead is shared between nomad_csi_volume and
+// nomad_csi_volume_registration because once a volume is stored in Nomad it
+// is read from the same endpoint, regardless of how it was created.
+func resourceCSIVolumeRead(d *schema.ResourceData, meta interface{}) error {
+	providerConfig := meta.(ProviderConfig)
+	client := providerConfig.client
+
+	id := d.Id()
+	opts := &api.QueryOptions{
+		Namespace: d.Get("namespace").(string),
+	}
+	if opts.Namespace == "" {
+		opts.Namespace = "default"
+	}
+	log.Printf("[DEBUG] reading information for CSI volume %q in namespace %q", id, opts.Namespace)
+	volume, _, err := client.CSIVolumes().Info(id, opts)
+	if err != nil {
+		// As of Nomad 0.4.1, the API client returns an error for 404
+		// rather than a nil result, so we must check this way.
+		if strings.Contains(err.Error(), "404") {
+			log.Printf("[DEBUG] CSI volume %q does not exist, so removing", id)
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("error checking for CSI volume: %s", err)
+	}
+	log.Printf("[DEBUG] found CSI volume %q in namespace %q", volume.Name, volume.Namespace)
+
+	d.Set("name", volume.Name)
+	d.Set("namespace", volume.Namespace)
+	d.Set("volume_id", volume.ID)
+	d.Set("external_id", volume.ExternalID)
```

Comment:
Moved the read logic to this resource since it seemed to make more sense name-wise. This was the only change.